### PR TITLE
DEVX-265: Moving code from quickstart-demos to here for microservices demo

### PR DESCRIPTION
**Pull request by ybyzek**

### CODE REVIEW

**Comment by vvcephei** [comment]
Hi @ybyzek ,

Now I see what you were getting at before. I think this will be a nice addition.

Do you mind adding a section to the Readme explaining how each of the scripts should be used?

Thanks,
-John

**Comment by ybyzek** [comment]
@vvcephei : do you mind if I defer that for now?  Two reasons:
1. There's existing high-level explanation at https://github.com/confluentinc/kafka-streams-examples/blob/5.0.0-post/src/main/java/io/confluent/examples/streams/microservices/README.md
2. I'm working on Docker-izing the demo along with a step-by-step self-paced tutorial, and as I flesh out ideas then it will be documented much better.

**Comment by vvcephei** [comment]
Sure, I guess that's fine as long as we eventually add something to the readme, either that tutorial, or a reference to it.

**Comment by vvcephei** [comment]
In the absence of that information, though, it's a little hard to review this PR. It takes a careful reading and some guesswork to infer from those classes how this tutorial would work, and therefore, whether we'd recommend any changes.

Does that make sense?

File: src/main/java/io/confluent/examples/streams/microservices/AddInventory.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.streams.StreamsConfig;
+
+import org.apache.kafka.common.serialization.Serdes;
+
+import org.glassfish.jersey.client.ClientConfig;
```
Comment (vvcephei) [inline src/main/java/io/confluent/examples/streams/microservices/AddInventory.java]: There are some unused imports in this class.

File: src/main/java/io/confluent/examples/streams/microservices/AddInventory.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    MonitoringInterceptorUtils.maybeConfigureInterceptorsProducer(producerConfig);
+    ProductTypeSerde productSerde = new ProductTypeSerde();
+
+    try (KafkaProducer<Product, Integer> stockProducer = new KafkaProducer<>(
+        producerConfig,
+        productSerde.serializer(),
+        Serdes.Integer().serializer()))
+    {
+      for (KeyValue<Product, Integer> kv : inventory) {
+        stockProducer.send(new ProducerRecord<>("warehouse-inventory", kv.key, kv.value))
```
Comment (vvcephei) [inline src/main/java/io/confluent/examples/streams/microservices/AddInventory.java]: Didn't use the constant topic name here.

File: src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  public static void main(String [] args) throws Exception {
+
+    final int NUM_CUSTOMERS = 6;
+    final List<Product> productTypeList = Arrays.asList(Product.JUMPERS, Product.UNDERPANTS, Product.STOCKINGS);
+    final Random randomGenerator = new Random();
+
+    final int restPort = args.length > 0 ? Integer.valueOf(args[0]) : 5432;
+    System.out.printf("restPort: %d\n", restPort);
```
Comment (vvcephei) [inline src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java]: `\n` should be `%n` (also elsewhere)

File: src/main/java/io/confluent/examples/streams/microservices/AddInventory.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    final Properties producerConfig = new Properties();
+    producerConfig.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+    producerConfig.put(ProducerConfig.ACKS_CONFIG, "all");
+    producerConfig.put(ProducerConfig.RETRIES_CONFIG, 0);
+    producerConfig.put(ProducerConfig.CLIENT_ID_CONFIG, "inventory-generator");
+    MonitoringInterceptorUtils.maybeConfigureInterceptorsProducer(producerConfig);
+    final ProductTypeSerde productSerde = new ProductTypeSerde();
+
+    try (KafkaProducer<Product, Integer> stockProducer = new KafkaProducer<>(
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/AddInventory.java]: nit: final

File: src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        System.out.printf("Posted order %d does not equal returned order: %s\n", i, returnedOrder.toString());
+      } else {
+        System.out.printf("Posted order %d equals returned order: %s\n", i, returnedOrder.toString());
+      }
+
+      // Send payment
+      final Payment payment = new Payment("Payment:1234", id(i), "CZK", 1000.00d);
+      sendPayment(payment.getId(), payment, Topics.PAYMENTS);
+  
+      // GET order, assert that it is Validated
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java]: Can the comments be removed?

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumeCustomers.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, "test-customers");
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        props.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "1000");
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://localhost:8081");
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, Topics.CUSTOMERS.keySerde().deserializer().getClass());
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, Topics.CUSTOMERS.valueSerde().deserializer().getClass());
+        MonitoringInterceptorUtils.maybeConfigureInterceptorsConsumer(props);
+
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/util/ConsumeCustomers.java]: nit: final

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumeOrders.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, "test-orders");
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        props.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "1000");
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://localhost:8081");
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, Topics.ORDERS.keySerde().deserializer().getClass());
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, Topics.ORDERS.valueSerde().deserializer().getClass());
+        MonitoringInterceptorUtils.maybeConfigureInterceptorsConsumer(props);
+
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/util/ConsumeOrders.java]: nit: final

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumePayments.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, "test-payments");
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        props.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "1000");
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://localhost:8081");
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, Topics.PAYMENTS.keySerde().deserializer().getClass());
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, Topics.PAYMENTS.valueSerde().deserializer().getClass());
+        MonitoringInterceptorUtils.maybeConfigureInterceptorsConsumer(props);
+
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/util/ConsumePayments.java]: nit: final

File: src/main/java/io/confluent/examples/streams/microservices/util/ProduceOrders.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        mySerializer.configure(
+            Collections.singletonMap(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://localhost:8081"),
+            isKeySerde);
+
+        final Properties props = new Properties();
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        props.put(ProducerConfig.ACKS_CONFIG, "all");
+        props.put(ProducerConfig.RETRIES_CONFIG, 0);
+        MonitoringInterceptorUtils.maybeConfigureInterceptorsProducer(props);
+        final KafkaProducer<String, Order> producer = new KafkaProducer<String, Order>(props, new StringSerializer(), mySerializer);
```
Comment (bbejeck) [inline src/main/java/io/confluent/examples/streams/microservices/util/ProduceOrders.java]: use try-with-resources for `Producer` like you did in `AddInventory` here and below.

File: src/main/java/io/confluent/examples/streams/microservices/AddInventory.java (hunk 1 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/AddInventory.java]: @vvcephei : I can't get `checkstyle` working to identify the unused imports in this class.  Do you mind telling which ones they are and then I'll remove them?

File: src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java (hunk 3 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java]: Ack

File: src/main/java/io/confluent/examples/streams/microservices/AddInventory.java (hunk 4 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/AddInventory.java]: Ack

File: src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java (hunk 5 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/PostOrdersAndPayments.java]: I removed some, not all

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumeCustomers.java (hunk 6 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/util/ConsumeCustomers.java]: Ack

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumeOrders.java (hunk 7 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/util/ConsumeOrders.java]: Ack

File: src/main/java/io/confluent/examples/streams/microservices/util/ConsumePayments.java (hunk 8 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/util/ConsumePayments.java]: Ack

File: src/main/java/io/confluent/examples/streams/microservices/util/ProduceOrders.java (hunk 9 above)
Comment (ybyzek) [reply src/main/java/io/confluent/examples/streams/microservices/util/ProduceOrders.java]: Ack

**Comment by vvcephei** [comment]
Hey @ybyzek , I've sent https://github.com/confluentinc/kafka-streams-examples/pull/166 to remove the unused imports. While I was at it, I also proposed to clean up some other stuff, like unused variables.

Feel free to take some or all of the changes, I felt that the PR was just the most efficient way to share the information with you; no pressure.

FWIW, the pr is targeting your branch, so if you want, you can just hit merge on #166 , and it will become part of #165 .

**Comment by ybyzek** [comment]
@vvcephei amazing, thank you for #166 :)  

May I now merge this PR?

**Pull request merged**